west: west twister integration

asemjonovs opened the pull request:
**West/Twister integration**

When multiple `zephyr` projects are in different places, calling `twister` may use different paths. This could lead to usage of mismatched version of twister with the current project in use.  See https://github.com/zephyrproject-rtos/zephyr/issues/50389 for more details.

**Design**
Integrate twister into west as an extension command `west twister`.  The path to twister shall be resolved in the following order

1. `ZEPHYR_BASE` environment variable
2. West config option `twister.path`
3. Current path of `zephyr_ext_common.py`

**Implementation**

- Refactor twister APIs to gain access to `parse_arguments` for new west extension can leverage same command line arguments.
- Create function to get twister path, which will resolve based on priority indicated in design above.
- Create `west twister` command extension

asemjonovs committed:
west: west twister integration

Add twister command to west.

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

hakehuang commented:
@asemjonovs can you update the west doc add a usage guide? besides, can we add a test for this?

asemjonovs committed:
twister: Split up argument parsing into smaller functions

Split argument parsing into smaller functions to be leveraged
by west twister extension.

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

asemjonovs committed:
west: Update documentation on twister script integration

Adding documentation for twister wrapper west command

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

asemjonovs committed:
ci: west_cmds: Add ply for twister support

Install ply for twister support

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

yperess commented:
Fixes #50389

carlescufi commented:
@mbolivar-nordic @nashif please take a look

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,82 @@
+"""
+
+
+def _banner(msg):
+    log.inf("-- west twister: " + msg, colorize=True)
+
```
Use west.log.banner instead.

asemjonovs replied:
done

mbolivar-nordic replied:
>> Use west.log.banner instead.
>
> done

What happened here? The banner went completely away.

asemjonovs replied:
Since I copy/pasted this from build.py, I realized that I didn't use this function within my class definition.  So I cleaned it up.

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,82 @@
+west twister [-h]
+"""
+
+BUILD_DESCRIPTION = f"""\
```
Unnecessary f-string; the linter may warn on this.

asemjonovs replied:
done

mbolivar-nordic commented on scripts/west_commands/zephyr_ext_common.py:
```diff
@@ -23,6 +26,30 @@
+    else:
+        twister_path = zephyr_base + "/scripts"
+
+    print(f"Using TWISTER_PATH={twister_path}")
```
Remove this

asemjonovs replied:
done

mbolivar-nordic commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -174,3 +174,20 @@ As does deleting them::
+.. _west-twister:
+
+Twister wrapper: ``west twister``
+*****************************************
```
Align `*` with the length of the section by removing extra characters.

asemjonovs replied:
done

mbolivar-nordic commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -174,3 +174,20 @@ As does deleting them::
+*****************************************
+This command is a wrapper for :ref:`twister <twister_script>`.  It leverages
+current :envvar:`ZEPHYR_BASE` environment variable and west manifest to launch
+the twister version associated with the west workspace. `west twister` resolves
```
You need two backticks for inline code in .rst. The single backtick syntax is markdown.

asemjonovs replied:
done

mbolivar-nordic commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -174,3 +174,20 @@ As does deleting them::
+This command is a wrapper for :ref:`twister <twister_script>`.  It leverages
+current :envvar:`ZEPHYR_BASE` environment variable and west manifest to launch
+the twister version associated with the west workspace. `west twister` resolves
+path to twister in the following order
+
+#. :envvar:`ZEPHYR_BASE` environment variable
+#. west manifest config for twister.path (see :ref:`west-config-cmd` for more details)
+#. Current path of zephyr_ext_common.py
+
+To set the path of twister for current west workspace use::
+
+  west config twister.path <path_to_twister>
```
Actually, thinking about this a bit more, is it really worth it to have a configuration option for the location of twister? If we don't have a use case for it already I would rather not add it.

asemjonovs replied:
I initially tried using config option `zephyr.base` for this but when testing scenarios where this didn't exist yet, lead it to be automatically populated with 
```
[zephyr]
base = zephyr
```
So on my second call when testing `west twister` used the wrong version of twister. I couldn't find the code that was doing this, so I went with the more direct and unused `twister.path` option name.  
My preference to give this as an option to the developer is so that they don't have to keep updating `ZEPHYR_BASE` environment variable, when switching between projects.  If this is configured in their west workspace, just changing to the directory will update them to the right path for twister.  I'm open to other ideas to achieve this behavior.

mbolivar-nordic replied:
> My preference to give this as an option to the developer is so that they don't have to keep updating ZEPHYR_BASE environment variable, when switching between projects.

What ]application types](https://docs.zephyrproject.org/latest/develop/application/index.html#application-types) are you referring to in this scenario? I assume by "project" you mean "application". 

If you have multiple zephyr repository applications or zephyr workspace applications, there is no need to set ZEPHYR_BASE at all when building within the workspace, for example. (I'm assuming you build within the workspace directories in this case.)

If you have multiple freestanding applications that need to use different west workspaces, there is no avoiding resetting ZEPHYR_BASE when switching between them to build, is there? (I'm assuming you build within the application directory in this case.)

> If this is configured in their west workspace, just changing to the directory will update them to the right path for twister.

Assuming "the directory" corresponds to my above, I see no need for twister.path.

> I'm open to other ideas to achieve this behavior.

Seems like you can find twister relative to zephyr_ext_common's ZEPHYR_SCRIPTS without needing the config option.

mbolivar-nordic replied:
> This would be a [freestanding](https://docs.zephyrproject.org/latest/develop/application/index.html#zephyr-freestanding-application) application type. With a directory structure something like

What you have shown is not a freestanding application. A freestanding application is outside of any west workspace. You have shown app1 and app2 directories that are in fact the topdirs of independent west workspaces.

mbolivar-nordic reviewed (changes requested):
Thanks for the patches!

mbolivar-nordic reviewed (changes requested):
Second round of review after having had some more time to think about this. These -1 comments are in addition to my previous lower-level ones.

- The pull request description needs to be updated to describe the use cases, design, and implementation if you want this to be merged and it is not just an "experiment"
- The documentation needs to be expanded to reflect the updated description. At least one example must be added for each of the use cases you are addressing.

asemjonovs commented on scripts/west_commands/zephyr_ext_common.py:
```diff
@@ -23,6 +26,30 @@
+    zephyr_base = os.environ.get("ZEPHYR_BASE")
+
+    if zephyr_base is None:
+        try:
+            topdir = Path(util.west_topdir(start=Path.cwd(), fall_back=True))
+            config = Configuration(topdir=topdir)
+            twister_path = config.get("twister.path", str(ZEPHYR_SCRIPTS))
```
I don't think self.config is safe for me to use in `do_add_parser`.  I've tried creating an `_import_twister` function to handle the dynamic loading of twister to use self.config in `do_add_parser` but I was getting errors.  I don't consider myself as a python expert, so if you have any tips for me here, please let me know :smiley:.  You may have noticed that I used `west_commands/build.py` as a template for my changes.

mbolivar-nordic replied:
The more I think about it, the less I like the idea of making this configurable. The twister_cmd.py and twisterlib implementations are in the same repository, and they're coupled. Getting twister_cmd.py from zephyr repository A and twisterlib from repository B would be like getting `west build` from repository A and the build system from repository B in some sense -- it just doesn't make sense to me and I think the case analysis by application type shows it's not necessary.

mbolivar-nordic replied:
... you still haven't addressed this; why are you asking for a new review?

asemjonovs replied:
Sorry - missed this request.  Removed west config for path.  Now it will check `ZEPHYR_BASE` and if that doesn't exist will base it off current path of `zephyr_ext_common.py`

asemjonovs commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -34,7 +34,7 @@
 
 def add_parse_arguments(parser = None):
-    if(parser is None):
+    if parser is None:
```
done

mbolivar-nordic replied:
> Unrelated change; please move to another commit.

Actually, looking at this again, this is a fixup commit -- the linter complaint was originally added in ` twister: Split up argument parsing into smaller functions`. I think what might have happened is you squashed it into the wrong commit when rebasing originally. At any rate it should be squashed into that first patch in the PR instead.

asemjonovs commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,82 @@
+            help=self.help,
+            formatter_class=argparse.RawDescriptionHelpFormatter,
+            description=self.description,
+            usage=BUILD_USAGE,
```
done

asemjonovs commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -174,3 +174,20 @@ As does deleting them::
+Twister wrapper: ``west twister``
+*****************************************
+This command is a wrapper for :ref:`twister <twister_script>`.  It leverages
+current :envvar:`ZEPHYR_BASE` environment variable and west manifest to launch
```
done

mbolivar-nordic replied:
Unless there's a really compelling reason for making this separately configurable (even via ZEPHYR_BASE) that I'm missing, I'm reopening this comment to update the docs to just say that it uses the twister that's in the same repository as the twister command itself.

asemjonovs replied:
I was giving capability to developer to be able to point to a different ZEPHYR_BASE, but I can see that also leading to problems.  I'll remove this capability.

mbolivar-nordic replied:
> I was giving capability to developer to be able to point to a different ZEPHYR_BASE, but I can see that also leading to problems.

The developer already has this capability though, is the thing. West respects ZEPHYR_BASE when looking for the workspace, so by the time the extension runs, the workspace has already been located in a way that respects this environment variable.  There doesn't seem to be a need for extra work here, because what can happen is that if you have `zephyr.base-prefer` set to `configfile` in your `.west/config` (see https://docs.zephyrproject.org/latest/develop/west/config.html#built-in-configuration-options), but ZEPHYR_BASE in the environment set to something else, you'll end up with a `west twister` implementation in workspace 1 that calls out to a twister implementation in workspace 2. It really doesn't make sense to me to do that because the twister APIs that the west extension is using are tightly coupled to what is in the current repository, and using another repository could break that. It's a footgun.

> I'll remove this capability.

Sounds good, thakns.

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,72 @@
+Convenience wrapper for twister.
+"""
+
+class Twister(Forceable):
```
I went back and double checked since you said you were using `west build` as inspiration. I'm guessing that's why this line is here. Inheriting from `Forceable` is not necessary for `west twister` because you do not ever call `check_force`. Just inherit from `WestCommand` instead, as recommended in the documentation:

 https://docs.zephyrproject.org/latest/develop/west/extensions.html#adding-a-west-extension.

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,72 @@
+from twisterlib.environment import add_parse_arguments, parse_arguments
+
+BUILD_DESCRIPTION = """\
+Convenience wrapper for twister.
+"""
```
I got around to trying this out today. The help output is much improved since not overriding the default, but I think we can do a little better here:

```
"""\
Convenience wrapper for twister. The below options are shared
with the twister script and have the same effects as if you ran twister
directly. Refer to the twister documentation for more information.
"""

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,72 @@
+sys.path.insert(0, str(twister_path))
+sys.path.insert(0, str(twister_path / "pylib" / "twister"))
+
+from importlib.util import spec_from_loader, module_from_spec
+from importlib.machinery import SourceFileLoader
+
+spec = spec_from_loader(
+    "twister", SourceFileLoader("twister", str(twister_path / "twister"))
+)
+twister = module_from_spec(spec)
+spec.loader.exec_module(twister)
+
+from twisterlib.environment import add_parse_arguments, parse_arguments
```
Hmm, giving this another look, why are you both using importlib **and** importing from twisterlib?

The importlib library is "heavy machinery" in my opinion and I think this may be a code smell.

Why can't the twister script be refactored a bit more so that the shared code you need can be imported from twisterlib using an ordinary import, instead of breaking out importlib?

asemjonovs replied:
I tried `import twister` but since `scripts/twister` is missing the `*.py` extension, it errors when normally imported.  I can move the entire `main` definition to another file to avoid this?  I initially created `twister.py` soft link to `twister` but this failed compliance checks.  Searching the web lead me to this solution to import files that don't have the python extension.

mbolivar-nordic replied:
> I can move the entire main definition to another file to avoid this?

Right, that is what I am saying -- move the necessary shared code to a python file inside twisterlib, and import that from both `twister` and `twister_cmd.py`.

mbolivar-nordic replied:
(This would need approval from the twister maintainers, but it seems like the simplest way to handle this.)

mbolivar-nordic reviewed (changes requested):
@asemjonovs responded to your comments; thanks for the updates.

I assume you still will update the documentation to include examples as requested?

asemjonovs commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,72 @@
+        options = self._parse_arguments(args=remainder, options=args)
+        twister.main(options)
+
+    def _add_parse_arguments(self, parser):
+        return add_parse_arguments(parser)
```
I added this to aid in writing my tests for this class.  I'll work on inlining this.

asemjonovs commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,72 @@
+    def _add_parse_arguments(self, parser):
+        return add_parse_arguments(parser)
+
+    def _parse_arguments(self, args, options):
+        return parse_arguments(self.parser, args, options)
```
I added this to aid in writing my tests for this class.  I'll work on inlining this.

mbolivar-nordic replied:
No need to inline this if you're using it in white box tests. Thanks for adding a comment to make it clear that that's what's going on.

asemjonovs commented:
> @asemjonovs responded to your comments; thanks for the updates.
> 
> I assume you still will update the documentation to include examples as requested?
@mbolivar-nordic - yeah, I'll update the documentation with examples.  Been bouncing between multiple issues today.

mbolivar-nordic reviewed (changes requested):
Here is another review. The remaining unresolved threads we still need to settle are:

1. Whether this should be configurable. I think you have misunderstood how freestanding applications are supposed to work. You need to use ZEPHYR_BASE in the environment to build freestanding applications. Freestanding applications are not west workspace topdirs, so the use case you've shown is not how the build system and west build are designed to work.
2. Documentation updates
3. Dropping the importlib heavy machinery and just using regular `import` for all the code you need to get from twister
4. Nitpick about the linter fix

asemjonovs committed:
twister: move main to twisterlib/twister_main.py

This allows for ability to import main to be used
within west

Signed-off-by: Al Semjonovs <asemjonovs@google.com>

yperess commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -174,3 +174,25 @@ As does deleting them::
+
+.. _west-twister:
+
+Twister wrapper: ``west twister``
```
Since people can override this, can we just call it `west test`. This way if someone wants to replace the test runner they can override the west command and call into their own. We can just document that by default the test runner is twister.

mbolivar-nordic replied:
> Since people can override this, can we just call it west test.

Wait, what? No, this is always going to use twister; this is not overridable

yperess replied:
ACK

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,63 @@
+from twisterlib.environment import add_parse_arguments, parse_arguments
+from twisterlib.twister_main import main
+
+BUILD_DESCRIPTION = """\
```
Nitpick: the variable name here is copypasta

mbolivar-nordic commented on scripts/west_commands/twister_cmd.py:
```diff
@@ -0,0 +1,63 @@
+from zephyr_ext_common import get_twister_path
+
+# Resolve path to twister libs and add imports
+twister_path = get_twister_path()
+os.environ["ZEPHYR_BASE"] = str(twister_path.parent)
```
Reminder to address this comment: https://github.com/zephyrproject-rtos/zephyr/pull/50601#discussion_r990503215

asemjonovs commented on scripts/west_commands/zephyr_ext_common.py:
```diff
@@ -23,6 +23,22 @@
 ZEPHYR_SCRIPTS = ZEPHYR_BASE / 'scripts'
 ZEPHYR_CMAKE = ZEPHYR_BASE / 'cmake'
 
+def get_twister_path():
+    """Get twister path using based on the following hierarchy
+    1) ZEPHYR_BASE env variable is set
+    2) Based on current path of this file.
+    """
```
After cleaning up abilities to configure this path, this function is no longer needed. Will remove this in upcoming changes.

asemjonovs commented on scripts/west-commands.yml:
```diff
@@ -15,6 +15,11 @@ west-commands:
       - name: build
         class: Build
         help: compile a Zephyr application
+  - file: scripts/west_commands/twister_cmd.py
```
Good catch! Thanks

hakehuang commented on scripts/twister:
```diff
@@ -195,225 +190,15 @@ if not ZEPHYR_BASE:
-        return 1
-
-    return 0
-
```
why removing this code?

asemjonovs replied:
@hakehuang - didn't remove, I created a function for this code which moved here
scripts/pylib/twister/twisterlib/twister_main.py

I broke this into smaller functions to be able to run twister from west.
